### PR TITLE
Acquire an token with the certificate of service principal

### PR DESCRIPTION
**Pull request by hellokangning**
**Reason for Change**:

Currently, Pod Identity supports user-assigned Managed Identity and the client scecret of service principal. For a higher level of assurance, Azure recommend using a certificate (instead of a client secret) as a credential.

**Issue Fixed**:

Fixes [#516](https://github.com/Azure/aad-pod-identity/issues/516)

### CODE REVIEW

File: pkg/nmi/managed.go [hunk 1]
```diff
@@ -70,8 +70,10 @@ func (mc *ManagedClient) GetToken(ctx context.Context, rqClientID, rqResource st
 		token, err := auth.GetServicePrincipalTokenFromMSIWithUserAssignedID(clientID, rqResource)
 		return token, err
 	case aadpodid.ServicePrincipal:
-		tenantid := azureID.Spec.TenantID
-		klog.Infof("matched identityType:%v tenantid:%s clientid:%s resource:%s", idType, tenantid, utils.RedactClientID(clientID), rqResource)
+		tenantID := azureID.Spec.TenantID
+		adEndpoint := azureID.Spec.ADEndpoint
+		klog.Infof("matched identityType:%v adendpoint:%s tenantid:%s clientid:%s resource:%s",
+			adEndpoint, idType, tenantID, utils.RedactClientID(clientID), rqResource)
```
Comment (chewong) [inline pkg/nmi/managed.go:76]: `adEndpoint` and `idType` should be swapped.

**Comment by aramase** [comment]
@hellokangning When you get a chance, could you rebase the PR and fix the missing arguments?

**Comment by hellokangning** [comment]
> @hellokangning When you get a chance, could you rebase the PR and fix the missing arguments?

Sorry for the late response, updated and re-run `go test`

**Comment by hellokangning** [comment]
Hey @chewong , it seems the unit test failed, but I don't the reason, could you take a look at why nmi build failed, I can run `make build` locally
```
2020-05-29T15:42:17.1509542Z FAIL	github.com/Azure/aad-pod-identity/pkg/nmi [build failed]
```

File: pkg/nmi/managed.go [hunk 2]
```diff
@@ -87,7 +89,20 @@ func (mc *ManagedClient) GetToken(ctx context.Context, rqClientID, rqResource st
 			clientSecret = string(v)
 			break
 		}
-		token, err := auth.GetServicePrincipalToken(tenantid, clientID, clientSecret, rqResource)
+		token, err := auth.GetServicePrincipalToken(adEndpoint, tenantID, clientID, clientSecret, rqResource)
+		return token, err
+	case aadpodid.ServicePrincipalCertificate:
+		tenantID := azureID.Spec.TenantID
+		adEndpoint := azureID.Spec.ADEndpoint
+		klog.Infof("matched identityType:%v adendpoint:%s tenantid:%s clientid:%s resource:%s",
```
Comment (aramase) [inline pkg/nmi/managed.go:97]: The error is here - 
```
Infof format %s has arg idType of wrong type github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity.IdentityType
```

Please change the order of args - "idType, adEndpoint tenantID, utils.RedactClientID(clientID), rqResource"

File: pkg/nmi/managed.go [hunk 3]
```diff
@@ -87,7 +89,20 @@ func (mc *ManagedClient) GetToken(ctx context.Context, rqClientID, rqResource st
...
+			adEndpoint, idType, tenantID, utils.RedactClientID(clientID), rqResource)
```
Comment (aramase) [inline pkg/nmi/managed.go:98]:
```suggestion
		klog.Infof("matched identityType:%v adendpoint:%s tenantid:%s clientid:%s resource:%s",
			idType, adEndpoint, tenantID, utils.RedactClientID(clientID), rqResource)
```

File: pkg/nmi/managed.go (hunk 2 above)
Comment (hellokangning) [reply pkg/nmi/managed.go:97]: Got it, thanks Anish

**Comment by chewong** [comment]
Hi @hellokangning, could you also fix the linting problem by running `make lint`? Thanks.

**Comment by hellokangning** [comment]
> Hi @hellokangning, could you also fix the linting problem by running `make lint`? Thanks.

Updated, but got no idea to fix "file not `goimports`-ed" after googling for a while, any suggestions?
```
pkg/auth/auth.go:7: File is not `goimports`-ed with -local github.com/Azure/aad-pod-identity (goimports)
```

**Comment by hellokangning** [comment]
> > Hi @hellokangning, could you also fix the linting problem by running `make lint`? Thanks.
> 
> Updated, but got no idea to fix "file not `goimports`-ed" after googling for a while, any suggestions?
> 
> ```
> pkg/auth/auth.go:7: File is not `goimports`-ed with -local github.com/Azure/aad-pod-identity (goimports)
> ```

Ignore this, fixed by adding a newline

**Comment by hellokangning** [comment]
@chewong  Hi Ernest, since all checks passed, could you merge this PR ?

File: pkg/nmi/standard_test.go [hunk 4]
```diff
@@ -71,6 +71,45 @@ func TestGetTokenForMatchingIDBySP(t *testing.T) {
...
+	fakeClient := fake.NewSimpleClientset()
+	reporter, err := metrics.NewReporter()
+	if err != nil {
+		t.Fatalf("expected nil error, got: %+v", err)
+	}
+	auth.InitReporter(reporter)
+
+	secret := &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "certificate"}, Data: make(map[string][]byte)}
+	val, _ := base64.StdEncoding.DecodeString("YWJjZA==")
+	secret.Data["Certificate"] = val
```
Comment (chewong) [inline pkg/nmi/standard_test.go:84]: s/Certificate/certificate

File: pkg/nmi/standard_test.go [hunk 5]
```diff
@@ -71,6 +71,45 @@ func TestGetTokenForMatchingIDBySP(t *testing.T) {
...
+	secret.Data["Password"] = val
```
Comment (chewong) [inline pkg/nmi/standard_test.go:85]: s/Password/password

File: deploy/demo/aadpodidentity-sp-cert.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+---
+apiVersion: "aadpodidentity.k8s.io/v1"
+kind: AzureIdentity
+metadata:
+  name: demo-aad1
+spec:
+  type: 2
+  tenantID: TENANT_ID
+  clientID: CLIENT_ID
+  clientPassword: {"Name":"demo-aad1-sp","Namespace":"default"}
```
Comment (aramase) [inline deploy/demo/aadpodidentity-sp-cert.yaml:18]: nit: new line

File: pkg/auth/auth.go [hunk 7]
```diff
@@ -130,6 +137,53 @@ func GetServicePrincipalToken(tenantID, clientID, secret, resource string) (*ada
...
+	begin := time.Now()
+	var err error
+
+	defer func() {
+		if err != nil {
+			err = reporter.ReportIMDSOperationError(metrics.AdalTokenOperationName)
+			if err != nil {
+				klog.Warningf("Metrics reporter error: %+v", err)
+			}
+
```
Comment (aramase) [inline pkg/auth/auth.go:151]: nit: remove new line

File: pkg/auth/auth.go [hunk 8]
```diff
@@ -130,6 +137,53 @@ func GetServicePrincipalToken(tenantID, clientID, secret, resource string) (*ada
...
+		}
+	}()
+
+	activeDirectoryEndpoint := defaultActiveDirectoryEndpoint
+	if adEndpointFromSpec != "" {
+		activeDirectoryEndpoint = adEndpointFromSpec
+	}
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, fmt.Errorf("creating the OAuth config: %v", err)
```
Comment (aramase) [inline pkg/auth/auth.go:166]: Let's change this error message to `fmt.Errorf("failed to create oauth config, err: %v", err)`

File: pkg/auth/auth.go [hunk 9]
```diff
@@ -130,6 +137,53 @@ func GetServicePrincipalToken(tenantID, clientID, secret, resource string) (*ada
...
+	}
+
+	privateKey, cert, err := pkcs12.Decode(certificate, password)
+	if err != nil {
+		return nil, err
```
Comment (aramase) [inline pkg/auth/auth.go:171]: can we wrap this error - `failed to decode certificate, err: %v`

File: pkg/nmi/standard.go [hunk 10]
```diff
@@ -166,8 +166,9 @@ func (sc *StandardClient) GetToken(ctx context.Context, rqClientID, rqResource s
 		token, err := auth.GetServicePrincipalTokenFromMSIWithUserAssignedID(clientID, rqResource)
 		return token, err
 	case aadpodid.ServicePrincipal:
-		tenantid := azureID.Spec.TenantID
-		klog.Infof("matched identityType:%v tenantid:%s clientid:%s resource:%s", idType, tenantid, utils.RedactClientID(clientID), rqResource)
+		tenantID := azureID.Spec.TenantID
+		adEndpoint := azureID.Spec.ADEndpoint
+		klog.Infof("matched identityType:%v tenantid:%s clientid:%s resource:%s", idType, tenantID, utils.RedactClientID(clientID), rqResource)
```
Comment (aramase) [inline pkg/nmi/standard.go:171]:
```
		klog.Infof("matched identityType:%v adendpoint:%s tenantid:%s clientid:%s resource:%s",
			idType, adEndpoint, tenantID, utils.RedactClientID(clientID), rqResource)
```

File: pkg/nmi/standard.go [hunk 11]
```diff
@@ -177,7 +178,19 @@ func (sc *StandardClient) GetToken(ctx context.Context, rqClientID, rqResource s
 			clientSecret = string(v)
 			break
 		}
-		token, err := auth.GetServicePrincipalToken(tenantid, clientID, clientSecret, rqResource)
+		token, err := auth.GetServicePrincipalToken(adEndpoint, tenantID, clientID, clientSecret, rqResource)
+		return token, err
+	case aadpodid.ServicePrincipalCertificate:
+		tenantID := azureID.Spec.TenantID
+		adEndpoint := azureID.Spec.ADEndpoint
+		klog.Infof("matched identityType:%v tenantid:%s clientid:%s resource:%s", idType, tenantID, utils.RedactClientID(clientID), rqResource)
```
Comment (aramase) [inline pkg/nmi/standard.go:186]: add `adEndpoint` to the logs

**Comment by hellokangning** [comment]
> Just few small nits! Also, we'll need to add docs about the new type and how to use it.

Thanks Anish, updated with docs

**Pull request merged**